rcljava/src/main/java/org/ros2/rcljava/node/NodeImpl.java

```diff
@@ -392,103 +401,299 @@ public final String getName() {
+    }
+
+    for (ParameterVariant p : parameters) {
+      // We already dealt with unsetting parameters above, so skip them
+      if (p.getType() == ParameterType.PARAMETER_NOT_SET) {
+        continue;
+      }
+      this.parameters.get(p.getName()).parameter = p;
+    }
+
+    result.setSuccessful(true);
+    return result;
+  }
+
+  public rcl_interfaces.msg.SetParametersResult setParameter(ParameterVariant parameter) {
+    synchronized (parametersMutex) {
+      return this.internalSetParametersAtomically(java.util.Arrays.asList(new ParameterVariant[] {parameter}));
     }
   }
 
   public List<rcl_interfaces.msg.SetParametersResult> setParameters(
       List<ParameterVariant> parameters) {
     List<rcl_interfaces.msg.SetParametersResult> results =
         new ArrayList<rcl_interfaces.msg.SetParametersResult>();
-    for (ParameterVariant p : parameters) {
-      rcl_interfaces.msg.SetParametersResult result =
-          this.setParametersAtomically(java.util.Arrays.asList(new ParameterVariant[] {p}));
-      results.add(result);
+
+    synchronized (parametersMutex) {
+      for (ParameterVariant p : parameters) {
+        rcl_interfaces.msg.SetParametersResult result =
+            this.internalSetParametersAtomically(java.util.Arrays.asList(new ParameterVariant[] {p}));
+        results.add(result);
+      }
     }
+
     return results;
   }
 
   public rcl_interfaces.msg.SetParametersResult setParametersAtomically(
       List<ParameterVariant> parameters) {
-    synchronized (mutex) {
-      rcl_interfaces.msg.SetParametersResult result = new rcl_interfaces.msg.SetParametersResult();
-      for (ParameterVariant p : parameters) {
-        ParameterAndDescriptor pandd = new ParameterAndDescriptor();
-        pandd.parameter = p;
+    synchronized (parametersMutex) {
+      return internalSetParametersAtomically(parameters);
+    }
+  }
 
-        rcl_interfaces.msg.ParameterDescriptor parameterDescriptor =
-            new rcl_interfaces.msg.ParameterDescriptor();
-        parameterDescriptor.setName(p.getName());
-        parameterDescriptor.setType(p.getType().getValue());
+  public void addOnSetParametersCallback(ParameterCallback cb) {
+    synchronized (parameterCallbacksMutex) {
+      this.parameterCallbacks.add(0, cb);
+    }
+  }
 
-        this.parameters.put(p.getName(), pandd);
-      }
-      result.setSuccessful(true);
-      return result;
+  public void removeOnSetParametersCallback(ParameterCallback cb) {
+    synchronized (parameterCallbacksMutex) {
+      this.parameterCallbacks.remove(cb);
+    }
+  }
+
+  private rcl_interfaces.msg.ParameterDescriptor internalDescribeParameter(String name) {
+    if (this.parameters.containsKey(name)) {
+      return this.parameters.get(name).descriptor;
+    } else if (this.allowUndeclaredParameters) {
+      return new rcl_interfaces.msg.ParameterDescriptor().setName(name);
+    }
+
+    throw new IllegalArgumentException("Parameter is not declared");
```

Comment:
Add the name of the parameter to the exception message.

Reply:
36daa72840b68189f0c6d7d763440c2395ec0d46